Resource Server Configurer

jzheaux opened the pull request:
So far, this is a draft on what may be the Resource Server
Configuration DSL. Feedback is welcome!

Issue: spring-projects/spring-security#5226

jzheaux committed:
Resource Server Configurer

So far, this is a draft on what may be the Resource Server
Configuration DSL. Feedback is welcome!

Issue: spring-projects/spring-security#5226

jgrandja commented on oauth2/oauth2-resource-server/src/main/java/org/springframework/security/config/annotation/web/configurers/oauth2/resourceserver/OAuth2ResourceServerConfigurer.java:
```diff
@@ -0,0 +1,140 @@
+
+import org.springframework.security.config.annotation.web.HttpSecurityBuilder;
+import org.springframework.security.config.annotation.web.configurers.AbstractHttpConfigurer;
+import org.springframework.security.oauth2.jose.jws.JwsAlgorithms;
```
The `config` module should not depend on `jose` or `jwt` packages/modules.
Only the `oauth2-client` and `oauth2-resource-server` modules may include the `oauth2-jose` dep

jzheaux replied:
I have a concern about `JwtDecoder` that seems to be related to this comment as well. (See below)

jgrandja commented on oauth2/oauth2-resource-server/src/main/java/org/springframework/security/config/annotation/web/configurers/oauth2/resourceserver/OAuth2ResourceServerConfigurer.java:
```diff
@@ -0,0 +1,140 @@
+
+	public BearerConfigurer bearer() {}
+
+	public class BearerConfigurer {
```
Let's simplify this to `http.oauth2().resourceServer().bearerTokenResolver(BearerTokenResolver bearerTokenResolver)`

If the user wants to enable form body resolving than they can do this:

```
DefaultBearerTokenResolver resolver = new DefaultBearerTokenResolver();
resolver.setUseFormEncodedBodyParameter(true)

http.oauth2().resourceServer().bearerTokenResolver(resolver)
```

jzheaux replied:
Okay, sure.

jgrandja commented on oauth2/oauth2-resource-server/src/main/java/org/springframework/security/config/annotation/web/configurers/oauth2/resourceserver/OAuth2ResourceServerConfigurer.java:
```diff
@@ -0,0 +1,140 @@
+
+
+
+	public JwtConfigurer jwt() {}
```
The DSL should not depend on`JwtDecoder`. The only configuration required for the initial feature implementation is JWK Set Url. I don't think anything more than that is required at the moment.

jzheaux replied:
What are you thoughts about the code needing to provide a custom `JwtDecoder` implementation?

jgrandja replied:
An `AuthenticationProvider` may be configured by the user and associate it with a custom `JwtDecoder`. The `AuthenticationProvider` can than be registered as a `@Bean`.

Or we can allow the user to register a `JwtDecoder` as a `@Bean` and than the `SecurityConfigurer` can discover it and supply it to the `JwtAuthenticationProvider`.

There are likely other options we can leverage without coupling it to the DSL.

jgrandja commented on oauth2/oauth2-resource-server/src/main/java/org/springframework/security/config/annotation/web/configurers/oauth2/resourceserver/OAuth2ResourceServerConfigurer.java:
```diff
@@ -0,0 +1,140 @@
+		public OAuth2ResourceServerConfigurer<B> and() {}
+	}
+
+	public class JwtDecoderConfigurer {
```
This can be removed as we only need to configure the JWK Set Url at the moment.

jgrandja commented on oauth2/oauth2-resource-server/src/main/java/org/springframework/security/config/annotation/web/configurers/oauth2/resourceserver/OAuth2ResourceServerConfigurer.java:
```diff
@@ -0,0 +1,140 @@
+		public JwtConfigurer and() {}
+	}
+
+	public class JwtVerifierConfigurer {
```
Providing specific configuration points for claims can get fairly large given that there are quite a few claims defined in JWT, OIDC, etc... I believe we can provide something like this to simplify:

`http.oauth2().resourceServer().accessTokenVerifiers(OAuth2TokenVerifier... verifiers)`

This configuration point can also be reused for other Access Token formats, for example, Opaque (Introspection Endpoint).

jzheaux replied:
Sounds good, I'll just refactor the `verifier` method and push it up a level. Makes sense.

jgrandja commented on oauth2/oauth2-resource-server/src/main/java/org/springframework/security/config/annotation/web/configurers/oauth2/resourceserver/OAuth2ResourceServerConfigurer.java:
```diff
@@ -0,0 +1,140 @@
+	}
+
+	public class JwkSetConfigurer {
+		public RemoteJwkSetConfigurer remote(String location) {}
```
JWK Set URL is the only configuration we require as of now. We can provide more configuration later when we implement a new feature, for example, configuration of RSA PK.

jzheaux replied:
Okay. Btw, I think I misunderstood your request from earlier, then. Was writing out what I anticipated the configuration would be, including anticipating features we were targeting for this milestone. Happy to remove all of that though (including the opaque token configuration that is also in this PR).

jgrandja replied:
Yeah maybe I miscommunicated. We do have to anticipate what the DSL will look like for opaque token and configurable verification key configurations but we won't implement that now.

We just need to ensure the design of the DSL will accommodate this seamlessly when we do implement.

jzheaux committed:
Reclassify to Verifiers and Formats

First, it appears to me that OAuth2 Access Tokens can be expressed in
a number of formats. From the JWT spec:

JSON Web Token (JWT) is a compact claims format

And in the case of an opaque token, the format is "opaque" to us
until, perhaps, it is processed.

For now, the DSL supports the configuration of JWT and Opaque as
formats supported by this resource server instance.

Roughly speaking, each format configured implies an additional
AuthenticationProvider.

In configuring each format, the DSL supports custom ways to process
tokens of that format.

Second, verification strategies stand independent of the token's
original format. Once any claims have actually been made, then they
can be verified.

At the very least, a token's claims to its own validity are
confirmed. For example, a token's iat and exp. This is not
configurable.

Additionally, there are two additional standard verification steps,
decrypting the claims and verifying the attached signature.

The DSL, then, supports configuration of these two verification
steps, though they will not result in instances of
OAuth2AccessTokenVerifier.

The DSL also allows for the specification of custom verification on
top of basic verification.

Issue: spring-projects/spring-security#5226

jgrandja commented on oauth2/oauth2-resource-server/src/main/java/org/springframework/security/config/annotation/web/configurers/oauth2/resourceserver/OAuth2ResourceServerConfigurer.java:
```diff
@@ -0,0 +1,159 @@
+			return new AccessTokenVerifiersConfigurer();
+		}
+
+		public AccessTokenFormatsConfigurer formats() {
```
What is the purpose of `formats()`? I understand it's directly associated to the format of the access token. However, doesn't `accessToken().jwt()` imply the JWT access token format?

jzheaux replied:
To be honest, I like the symmetry. Here are the access token formats I support: --formats listed--, here are the access token verifiers I support: --verifiers listed--.

Rob makes a similar comment as you, and I'm happy to defer, especially because it more likely that both your suggestions will be in line with what has been done in the past.

I don't have any functional reason, I just think it reads better.

jgrandja replied:
See earlier comment

jgrandja commented on oauth2/oauth2-resource-server/src/main/java/org/springframework/security/config/annotation/web/configurers/oauth2/resourceserver/OAuth2ResourceServerConfigurer.java:
```diff
@@ -0,0 +1,159 @@
+			return new EncryptionVerificationConfigurer();
+		}
+
+		public AccessTokenVerifiersConfigurer addVerifier(OAuth2TokenVerifier verifier) {
```
It's not a common DSL naming convention to prefix with `add`

rwinch replied:
I'd stick w/ add since it is additive vs overriding a value

jgrandja replied:
I would also like us to consider:

`.verifiers(List<OAuth2TokenVerifier<T extends AbstractOAuth2Token>> verifiers)`

jgrandja commented on oauth2/oauth2-resource-server/src/main/java/org/springframework/security/config/annotation/web/configurers/oauth2/resourceserver/OAuth2ResourceServerConfigurer.java:
```diff
@@ -0,0 +1,159 @@
+		return this;
+	}
+
+	public AccessTokenConfigurer accessToken(OAuth2TokenVerifier... verifiers) {
```
I don't see this being used in your sample configurations above?

jzheaux replied:
You appear to be correct. That is a left-over from the last iteration. The intent is:

`public AccessTokenConfigurer accessToken()`

jgrandja commented on oauth2/oauth2-resource-server/src/main/java/org/springframework/security/config/annotation/web/configurers/oauth2/resourceserver/OAuth2ResourceServerConfigurer.java:
```diff
@@ -0,0 +1,159 @@
+		}
+	}
+
+	private interface OAuth2AccessTokenProcessor<T extends AbstractOAuth2Token> {
```
What is `OAuth2AccessTokenProcessor` used for?

jzheaux replied:
Was looking for language in the spec that indicates the work that is done going from a token to a set of claims. Wanted something that made the notion of `JwtDecoder` more generic.

We have this same kind of thing with opaque tokens: Take this opaque token, go to this endpoint, confirm its validity, and (maybe) turn it into a claim set.

Essentially processing is: Take this token and turn it into a claim set.

jgrandja replied:
It's still not clear to me why we need this. If the user provided this configuration:

`oauth2().resourceServer().accessToken().jwt().jwkSetUri(jwkSetUri)`

than the `SecurityConfigurer` would initialize a `JwtAccessTokenAuthenticationProvider` and associate it with a `NimbusJwtDecoderJwkSupport` based on this configuration.

The point is that the `JwtDecoder` implementation would be known to the `SecurityConfigurer` based on the configuration.

Make sense?

jzheaux replied:
Agreed that we could correctly infer the default configuration from what you have listed.

What I am looking for is the right way for code to specify overrides to the defaults. I remember that you mentioned this as a possibility:

```java
@Bean
public JwtDecoder jwtDecoder() {
    // ...
}

@Bean
public OpaqueTokenIntrospector inspector() {
    // ... or whatever we call the thing that translates an opaque token into an instance of AbstractOAuth2Token
}
```
Is that the approach you would advocate?

If so, I guess what I'm not clear on is why the above way is more desirable than this:

```java
oauth2().resourceServer().accessToken().jwt().processor(decoder());

// ... or

oauth2().resourceServer().accessToken()
    .opaque().processor(introspector()).verifiers(...).and()
    .jwt().processor(decoder()).verifiers(...);
```

To me, with the @Bean way, I'd need to look it up in the documentation whereas this way, I can do tab completion to see what configuration options I have. Here, I can read it like I read the spec "My Jwt Processor is X".

jgrandja commented on oauth2/oauth2-resource-server/src/main/java/org/springframework/security/config/annotation/web/configurers/oauth2/resourceserver/OAuth2ResourceServerConfigurer.java:
```diff
@@ -0,0 +1,159 @@
+		}
+	}
+
+	public class EncryptionVerificationConfigurer {
```
This initial feature will not provide support for JWE so let's leave this out.

jzheaux replied:
See my comment at the end of the review, would love to get your take, too.

jgrandja commented on oauth2/oauth2-resource-server/src/main/java/org/springframework/security/config/annotation/web/configurers/oauth2/resourceserver/OAuth2ResourceServerConfigurer.java:
```diff
@@ -0,0 +1,159 @@
+	}
+
+	public class SignatureVerificationConfigurer {
+		public SignatureVerificationConfigurer keys(String uri) {
```
What if the user wanted an option to configure a PEM encoded PK which is also a `String`. This naming would collide.

jzheaux replied:
Are you saying that folks may want to indicate the raw key as a string? I'd more expect to see a `byte[]` if I wanted to hand in a raw value.

We can certainly support that, though I was imagining handing in an instance of `Key`, `SecretKey`, or the like.

I'm hoping to overload the `keys` method instead of doing something like `keysUri`, `keysValue`, etc.. I think that having all of them have the same name communicates more clearly that you get to pick one (though I'm happy to be swayed).

So, what about:

```java
keys(URI location)
keys(byte[] encoded)
keys(Key key)
keys(Map<String, Key> keysByVersion)
```

Auth0 also has an interface that I like which we could pattern after:

```java
keys(KeyProvider provider)
```

jzheaux replied:
Of course, I think this refers to a ticket that we haven't actually implemented yet, and perhaps our thoughts will change once that is complete.

jgrandja replied:
> Are you saying that folks may want to indicate the raw key as a string? I'd more expect to see a byte[] if I wanted to hand in a raw value.

Yes, Spring Boot 1.5.x supports this - [see](https://docs.spring.io/spring-boot/docs/1.5.12.RELEASE/reference/htmlsingle/#boot-features-security-custom-user-info)

```
security:
    oauth2:
        resource:
            jwt:
                keyValue: |
                    -----BEGIN PUBLIC KEY-----
                    MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKC...
                    -----END PUBLIC KEY-----
```

jgrandja replied:
> We can certainly support that, though I was imagining handing in an instance of Key, SecretKey, or the like.

Agreed, I see us initially supporting `java.security.PublicKey` and `String` for the verification keys.

jgrandja replied:
> I'm hoping to overload the keys method

I would prefer the overloading as well but how does the user know they are configuring a verification key vs. a signer key? This needs to be clear to the user. Obviously `PublicKey` and `PrivateKey` would make it clear but than it would still be fuzzy when the user wanted to configure key encryption (for example, JWE).

`keys().key(java.security.PrivateKey)`

It's not clear if this is for configuring the Signing key or Decryption key?

We might need to be more explicit than general overloading, for example:

```
keys()
    .verifyKey(rsaKey1)		// type java.security.PublicKey - used to verify signature
	.decryptKey(rsaKey2)	// type java.security.PrivateKey - used to decrypt token content
```

Something to consider for the design...

jzheaux replied:
Fair. Good feedback.

jzheaux replied:
> how does the user know they are configuring a verification key vs. a signer key?

From context. If we are configuring a resource server, we need decryption and verification:

```java
verifiers()
    .signature().keys()
    .encryption().keys()
````

makes it clear that the keys after `signature` are for the signature (verification), for example.

jgrandja replied:
Yes it does make it clear but I'm not sure it makes sense to have the `keys()` defined under the `verifiers()`. An `OAuth2TokenVerifier` is responsible for validating the claims/attributes of the access token after the signature has been verified or content has been decrypted. Now I'm wondering if the name `OAuth2TokenVerifier` is misleading?

rwinch commented on oauth2/oauth2-resource-server/src/main/java/org/springframework/security/config/annotation/web/configurers/oauth2/resourceserver/OAuth2ResourceServerConfigurer.java:
```diff
@@ -0,0 +1,159 @@
+/**
+ * Example configuration:
+ *
+ * oauth2().resourceServer().accessToken()
```
I'd simplify the hierarchy some. 

```java
oauth2()
     .resourceServer()
         .jwt()
         .verifiers()
             ...
```

jgrandja replied:
Agreed, let's try to simplify the hierarchy. I will add to what @rwinch has proposed:

```
oauth2()
    .resourceServer()
		.accessToken()
		    .verifiers(verifiers)	// verifiers -> List<OAuth2TokenVerifier<T extends AbstractOAuth2Token>>
			.jwt()
			    .jwkSetUri(jwkSetUri)	// URI used to retrieve the JWK Set
			.opaque()
			    .introspectionUri(uri);	// URI for the Introspection Endpoint
            ...
```

I feel like we need that extra hierarchy level in `.accessToken()` to indicate we are configuring a specific Access Token formats.

For the feature (to be implemented) that will allow the user to configure a RSA Public Key we could have something like this:

```
oauth2()
    .resourceServer()
	    .keys()
		    .verifyKey(rsaKey1)		// type java.security.PublicKey - used to verify signature
			.decryptKey(rsaKey2)	// type java.security.PrivateKey - used to decrypt token content
            ...
```

The hierarchy under `.keys()`  can than be leveraged by both `jwt` and `opaque` - no need to define the PK twice. It's a valid use case that the Authorization Server supports both JWT and Opaque (Introspection Endpoint) token formats.

rwinch commented:
I'd make sure to stick w/ features we have now.

jzheaux commented:
Since this was released in Spring Security 5.1, there isn't much need at this point to keep this around.
